### ISSUE TITLE
Tensorflow unable to detect no. of CPU cores

### ISSUE THREAD

**Issue by npanpaliya** [tensorflow/tensorflow#3545]
### Environment info
**Operating System:**
Ubuntu 14.04 Linux ppc64le

**Installed version of CUDA and cuDNN: 7.5**
ls -l /usr/local/cuda-7.5/lib64/libcud*
-rw-r--r-- 1 root root   326744 Nov  9  2015 /usr/local/cuda-7.5/lib64/libcudadevrt.a
lrwxrwxrwx 1 root root       16 Nov  5  2015 /usr/local/cuda-7.5/lib64/libcudart.so -> libcudart.so.7.5
lrwxrwxrwx 1 root root       19 Nov  5  2015 /usr/local/cuda-7.5/lib64/libcudart.so.7.5 -> libcudart.so.7.5.23
-rwxr-xr-x 1 root root   445192 Nov  5  2015 /usr/local/cuda-7.5/lib64/libcudart.so.7.5.23
-rw-r--r-- 1 root root   902750 Nov  9  2015 /usr/local/cuda-7.5/lib64/libcudart_static.a
lrwxrwxrwx 1 root root       13 Apr 24 20:17 /usr/local/cuda-7.5/lib64/libcudnn.so -> libcudnn.so.5
-rwxr-xr-x 1 root root 60068392 Apr 22 19:18 /usr/local/cuda-7.5/lib64/libcudnn.so.5.0.5
-rw-r--r-- 1 root root 59436124 Apr 22 19:30 /usr/local/cuda-7.5/lib64/libcudnn_static.a
**

**Tensorflow 0.8 built and installed from source. (v0.8.0 tag)**
1. The commit hash (`git rev-parse HEAD`) : 4b7bc3174ed67b4a0eb1803537c9d00f132e9ae7
2. The output of `bazel version`: 0.2.0
Note: Same issue also exists with tensorflow 0.9

**Steps to reproduce:**
On python prompt, 
1. import tensorflow as tf
2. tf.Session()

In a large output of above statement, there is a line that says "can't determine no. of CPU cores, assuming 4".

After looking into the tensorflow code, I found that __linux macro in tensorflow/core/platform/posix/port.cc is found to be not defined and hence sched_getaffinity is not called, and returns 4 as default value. I checked that usually on linux, __linux is defined but if we compile using -std=c++11, then this macro gets undefined. 
Anyway, even if sched_getaffinity is called, it returns logical no. of CPU cores i.e. maximum threads possible. On POWER, usually this count is pretty higher and having so many threads degrades the performance. Ideally, we should have physical core size as the max possible thread count in thread pool for optimum performance. I tried looking for a built-in function that gives me physical core size, but unfortunately, all of the functions I found gives me logical core size. Then I found another library "[libhwloc](https://www.open-mpi.org/projects/hwloc/)" that gives C API to get this information. This library in general is very useful with nice documentation. It is also available on canonical and one can install it using "apt-get install libhwloc-dev" on Ubuntu system or on RHEL using "yum install hwloc-devel". 
So, I wanted to know if Tensorflow can accept this additional dependency on hwloc library.

**Comment by michaelisard** [comment]
@martinwicke can you comment on dependency questions?

**Comment by npanpaliya** [comment]
Hi, 
Could you please let me know if hwloc can be added in the dependency list of tensorflow?

**Comment by martinwicke** [comment]
We can add it as a source dependency. @michaelisard what's your take on how useful this is, as opposed to fixing the `sched_getaffinity` issue directly? I'd lean toward the latter to fix this bug.

**Comment by michaelisard** [comment]
@cwhipkey you have been looking at threadpools recently. What do you think?

**Comment by npanpaliya** [comment]
@cwhipkey could you please let us know your thoughts?

**Comment by cwhipkey** [comment]
I think it would be good to figure out the issue with __linux not being set properly.

Can you try:
`touch /tmp/foo.cc`
`cpp -dM -std=c++11 /tmp/foo.cc`

vs
`cpp -dM /tmp/foo.cc`

Save the output of each separately and attach both files?  It will be good to see which macros are actually defined in both cases.

**Comment by npanpaliya** [comment]
Here are the files with both outputs. The one with std=c++11 doesn't have __linux defined (it has __linux__ defined though) while the other one without it, has __linux and __linux__ defined.

Even if __linux is defined, sched_getaffinity does not give us expected output. It gives us logical no. of cores, i.e. max possible threads (no. of sockets * core per socket * threads per core) which is pretty high on Power and it degrades the performance adversely. We also tested that using physical core size gives the best performance. All the existing relevant APIs like std::thread::hardware_concurrency(), omp_get_max_threads(), omp_get_num_procs(), get_nprocs(), etc. give the logical core size. We could not find any API giving us required size. 
Hwloc library is the one that provides the required API. We would anyway call this API only for power architectures, for non-power arch, sched_getaffinity will still be functional.

**Comment by cwhipkey** [comment]
Thanks for posting that info.  We'll make a change to use __linux__ instead of __linux, so that would fix the issue with using the hardcoded value instead of sched_getaffinity.

For logical vs physical, it does seem useful to be able to detect physical and use on PPC (unless there is some other change that can be made to use the logical cores in a more useful way).  I don't have experience with powerpc, so I don't know why using the logical cores makes it worse - is there a particular kind of computation that gets much worse in that case?

**Comment by npanpaliya** [comment]
What I've observed on most of our development machines that no. of threads per core is usually 8. So, sched_getaffinity returns values like 160 or 128. This value is (no. of threads per core * no. of cores per socket * no. of sockets). We did perform various tests with different thread count, and with 160 threads, we got degraded performance, and with even 4, the performance was no good. So, trying out further values, we settled at 32 or 40 which gave us the best performance. This means threads equal to actual physical no. of cores will work the best. On x86, we usually have less no. of threads per core (probably 1 or 2) which will give thread count approximately 20 or 32 that must be the best bet. I'm not sure about exact computation that suffered the performance hit. We tested Googlenet-tensorflow Inception v3 models for training some images. I'll get back to you on this.

**Comment by npanpaliya** [comment]
We actually tested Googlenet-tensorflow Inception V3 models, wherein we tried training a very small set of imagenet dataset (around 20 images) for just 10 steps with 1,2 and 4 GPUs. And, we ran this for different core sizes.

**Comment by npanpaliya** [comment]
Hello,

Could you please let me know if I can go ahead with hwloc library for retrieving physical core size?

**Comment by cwhipkey** [comment]
I think it's okay to include that library for this purpose.

I am still uncertain why it would have so much impact on a benchmark that
uses GPUs for the core computation - the main CPU work would be just the
scheduling of GPU kernels.

Note that this recent commit would affect any GPU benchmark as well:
https://github.com/tensorflow/tensorflow/commit/f1acb3bd828a73b15670fc8019f06a5cd51bd564
.  It's possible that that change, which would reduce the amount of thread
switching when running a GPU benchmark, will also help your case.

**Comment by npanpaliya** [comment]
Thank you @cwhipkey . We'll test with this commit and let you know the status of whether we need to address core size problem using hwloc library.

**Comment by npanpaliya** [comment]
@cwhipkey , we tested the recent commit referred by you. With macro change from __linux to __linux__, that recent commit worked well. We see better performance with this commit which is equally comparable with what we got using physical core size. Hence, for power, what we need is 

1. Changing __linux to __linux__ in port.cc where we call get_schedaffinity so that the function is called even on power
2. This recent commit that reduces the amount of thread switching in GPU mode.

Thanks for making the 1st change in the repo and fixing the 2nd one. We can close the issue now.
